src/app/clusters/general-commissioning-server/general-commissioning-server.cpp

```diff
@@ -218,60 +373,90 @@ bool emberAfGeneralCommissioningClusterCommissioningCompleteCallback(
     auto & failSafe               = Server::GetInstance().GetFailSafeContext();
     auto & fabricTable            = Server::GetInstance().GetFabricTable();
 
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
     ChipLogProgress(FailSafe, "GeneralCommissioning: Received CommissioningComplete");
 
     Commands::CommissioningCompleteResponse::Type response;
+
+    // Fail-safe must be armed
     if (!failSafe.IsFailSafeArmed())
     {
         response.errorCode = CommissioningErrorEnum::kNoFailSafe;
+        commandObj->AddResponse(commandPath, response);
+        return true;
     }
-    else
+
+    SessionHandle handle = commandObj->GetExchangeContext()->GetSessionHandle();
+
+    // Ensure it's a valid CASE session
+    if ((handle->GetSessionType() != Session::SessionType::kSecure) ||
+        (handle->AsSecureSession()->GetSecureSessionType() != SecureSession::Type::kCASE) ||
+        (!failSafe.MatchesFabricIndex(commandObj->GetAccessingFabricIndex())))
+    {
+        response.errorCode = CommissioningErrorEnum::kInvalidAuthentication;
+        ChipLogError(FailSafe, "GeneralCommissioning: Got commissioning complete in invalid security context");
+        commandObj->AddResponse(commandPath, response);
+        return true;
+    }
+
+#if CHIP_CONFIG_TC_REQUIRED
+    TermsAndConditionsProvider * const termsAndConditionsProvider = Server::GetInstance().GetTermsAndConditionsProvider();
+
+    // Ensure required terms and conditions have been accepted, then attempt to commit
+    if (nullptr != termsAndConditionsProvider)
     {
-        SessionHandle handle = commandObj->GetExchangeContext()->GetSessionHandle();
-        // If not a CASE session, or the fabric does not match the fail-safe,
-        // error out.
-        if (handle->GetSessionType() != Session::SessionType::kSecure ||
-            handle->AsSecureSession()->GetSecureSessionType() != SecureSession::Type::kCASE ||
-            !failSafe.MatchesFabricIndex(commandObj->GetAccessingFabricIndex()))
+        response.errorCode = CheckTermsAndConditionsAcknowledgements(termsAndConditionsProvider);
+        if (CommissioningErrorEnum::kOk != response.errorCode)
         {
-            response.errorCode = CommissioningErrorEnum::kInvalidAuthentication;
-            ChipLogError(FailSafe, "GeneralCommissioning: Got commissioning complete in invalid security context");
+            ChipLogError(FailSafe, "GeneralCommissioning: Terms and conditions not accepted");
+            commandObj->AddResponse(commandPath, response);
+            return true;
         }
-        else
+
+        if (failSafe.UpdateTermsAndConditionsHasBeenInvoked())
```

Comment:
I don't see anything in the spec that says that terms and conditions bits are rolled back by fail-safe expiration, or indeed any interaction between fail-safe and terms and conditions bits.  Where is this coming from?

Reply:
Let's fix the spec. It was the intent that there would be fail-safe handling during commissioning fail-safe, AND it's implemented. I would be OK letting it stay like this until the spec is updated, assuming it's already the assumption.

Reply:
This has been discussed offline, but to add clarification here as well this was following the recommendation here: https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/69f5bba9e65c83fec4b886ebe509f3d01f37d09f/src/service_device_management/GeneralCommissioningCluster.adoc?plain=1#L360